### PR TITLE
Refactor mocha-find-project-root using f functions

### PR DESCRIPTION
**Pull request by rejeep**
Also write specs for it.

@scottaj I also added a spec for what you commented on in the other PR for which I originally did this change.

### CODE REVIEW

File: mocha.el [hunk 1]
```diff
@@ -71,11 +71,8 @@ From http://benhollis.net/blog/2015/12/20/nodejs-stack-traces-in-emacs-compilati
 
 (defun mocha-find-project-root ()
   "Find the root of the project."
-  (let ((root-files '("package.json" ".git" ".hg" ".svn")) (dir nil) (i 0))
-    (while (not dir)
-      (setq dir (locate-dominating-file default-directory (nth i root-files)))
-      (setq i (+ i 1)))
-    dir))
+  (let ((root-dir (f--traverse-upwards (f-exists? (f-expand "package.json" it)))))
```
Comment (scottaj) [inline mocha.el]: It's late and maybe I'm totally missing it, but in `(f-expand "package.json" it)` what is `it`

Comment (rejeep) [inline mocha.el]: It's an anaphoric function, which is a bit magic for sure. 😄 

It works basically like this:

```lisp
(defmacro with-it (&rest body)
  `(let ((it "the thing")) ,@body))

(with-it
 (print it)) ;; "the thing"
```

So the macro will expand into this:

```lisp
(let ((it "the thing")) (print it))
```

It's very handy. Take a look at [dash](https://github.com/magnars/dash.el) where it's used all over the place.

Comment (scottaj) [inline mocha.el]: Ah okay, so it is a literal symbol name it binds to instead of a gensym? That looks cool.

**Comment by scottaj** [comment]
This LGTM, but ended up with conflicts.

**Comment by rejeep** [comment]
Rebased.

**Pull request merged**